### PR TITLE
qemu_vm: modify/optimize the code to support the parse the block data of "luks" image format.

### PR DESCRIPTION
**Pull request by nickzhq**
qemu_vm: modify/optimize the code to support the parse the block data of "luks" image format.

Support the parse the the multi-nested json format of "luks" image format. The get_block_old() can support the both "luks" and "qcow2".

ID: 1993837

Signed-off-by: Houqi (Nick) Zuo <hzuo@redhat.com>

### CODE REVIEW

**Comment by nickzhq** [comment]
qcow2 format:   
(1/1) Host_RHEL.m9.u0.qcow2.virtio_blk.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.block_auto_detect_size_increased: STARTED
| (1/1) Host_RHEL.m9.u0.qcow2.virtio_blk.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.block_auto_detect_size_increased: PASS (564.74 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
  
  
luks format:  
(1/1) Host_RHEL.m9.u0.luks.virtio_blk.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.block_auto_detect_size_increased: STARTED
| (1/1) Host_RHEL.m9.u0.luks.virtio_blk.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.block_auto_detect_size_increased: PASS (576.72 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0

**Comment by nickzhq** [comment]
Hello @luckyh @zhencliu , please take a look, thanks.

**Comment by luckyh** [comment]
@YongxueHong would you mind to help review this PR? thanks in advance!

**Comment by YongxueHong** [comment]
Hi @nickzhq 
Would you like to adjust the commit message by the following?

```
Write a good commit message, pointing motivation, issues that you’re addressing. 
Usually you should try to explain 3 points in the commit message: motivation, approach and effects:

header                  <- Target name: header content.(limited to 72 characters. No period.)
                        <- Blank line
message                 <- Any number of lines, limited to 72 characters per line.
                        <- Blank line
Signed-off-by:          <- Signature.
```
Thanks.

File: virttest/qemu_vm.py [hunk 1]
```diff
@@ -4864,11 +4864,9 @@ def is_json_data(data):
                             v = parse_json(v)
                         if k != key:
                             continue
-                        if v != value:
-                            matched = False
-                        break
-                    else:
-                        matched = False
+                        if value in str(v):
```
Comment (zhencliu) [inline virttest/qemu_vm.py:4867]: value can be a dict , so we cannot simply check it in this way, even if value is always a string, we may also cannot check it in this way, take block_auto_detect_size_increased as an example:

p_dict = {'file': '/home/kvm_autotest_root/images/stg.luks'}
but the item in block here is 
{'file': {'driver': 'luks', 'file': {'driver': 'file', 'filename': '/home/kvm_autotest_root/images/stg.luks'}, 'key-secret': 'image1_encrypt0'}, so
k = 'file'
v =  {'driver': 'luks', 'file': {'driver': 'file', 'filename': '/home/kvm_autotest_root/images/stg.luks'}, 'key-secret': 'image1_encrypt0'}
of course here value('/home/kvm_autotest_root/images/stg.luks) is in str(v), but its key('file') is not 'filename'. 
For luks image, it's reasonable to set p_dict = {'filename':'  '/home/kvm_autotest_root/images/stg.luks'}, but in this way, k != key, so we need to work out this. Some thoughts:
1. p_dict is a filter, we use this filter to get the exact node-name
2. p_dict can be a sub-set of value with essential information, here we don't need to set p_dict =  {'file':{'driver': 'luks', 'file': {'driver': 'file', 'filename': '/home/kvm_autotest_root/images/stg.luks'}, 'key-secret': 'image1_encrypt0'} }, only set p_dict = {'filename': '/home/kvm_autotest_root/images/stg.luks'} can get the exact device node. Give another example to represent a dedicated iscsi image: {"portal": "xxx", "target":"xxx", "lun":"0", "driver": "iscsi"}, we don't care about others.
3. do we need to pass p_dict as a dict with nested dict obj? e.g. {xxx{}}
@YongxueHong @nickzhq

Comment (YongxueHong) [reply virttest/qemu_vm.py:4867]: > value can be a dict , so we cannot simply check it in this way, even if value is always a string, we may also cannot check it in this way, take block_auto_detect_size_increased as an example:
> 
> p_dict = {'file': '/home/kvm_autotest_root/images/stg.luks'} but the item in block here is {'file': {'driver': 'luks', 'file': {'driver': 'file', 'filename': '/home/kvm_autotest_root/images/stg.luks'}, 'key-secret': 'image1_encrypt0'}, so k = 'file' v = {'driver': 'luks', 'file': {'driver': 'file', 'filename': '/home/kvm_autotest_root/images/stg.luks'}, 'key-secret': 'image1_encrypt0'} of course here value('/home/kvm_autotest_root/images/stg.luks) is in str(v), but its key('file') is not 'filename'. For luks image, it's reasonable to set p_dict = {'filename':' '/home/kvm_autotest_root/images/stg.luks'}, but in this way, k != key, so we need to work out this. Some thoughts:
> 
>     1. p_dict is a filter, we use this filter to get the exact node-name
> 
>     2. p_dict can be a sub-set of value with essential information, here we don't need to set p_dict =  {'file':{'driver': 'luks', 'file': {'driver': 'file', 'filename': '/home/kvm_autotest_root/images/stg.luks'}, 'key-secret': 'image1_encrypt0'} }, only set p_dict = {'filename': '/home/kvm_autotest_root/images/stg.luks'} can get the exact device node. Give another example to represent a dedicated iscsi image: {"portal": "xxx", "target":"xxx", "lun":"0", "driver": "iscsi"}, we don't care about others.
> 
>     3. do we need to pass p_dict as a dict with nested dict obj? e.g. {xxx{}}
>        @YongxueHong @nickzhq

Hi @nickzhq 
I remembered that I fixed to support luks format issue by https://github.com/avocado-framework/avocado-vt/pull/2629 .

Hi, @zhencliu 
I agree with your solution by fixing it from the caller(test case level)
suggestion:
```
        if data_image.image_format == 'luks':
            filename = get_image_json(img_tag, params, root_dir)
        device_name = vm.get_block({"filename": filename})
```
Thanks.

Comment (zhencliu) [reply virttest/qemu_vm.py:4867]: Hi @YongxueHong , make sense, we can fix it in test case.

Another concern here, if we want to find an image with backing file, e.g. 
json:{"backing": {"driver": "qcow2", "file": {"driver": "file", "filename": "/home/kar/workspace/root/avocado/data/avocado-vt/data.qcow2"}}, "driver": "qcow2", "file": {"driver": "file", "filename": "/home/kar/workspace/root/avocado/data/avocado-vt/sn1.qcow2"}}  >>> J1

but the get_image_json will not return the json string with the backing info, it just return the following:
json:{"file": {"driver": "file", "filename": "/home/kar/workspace/root/avocado/data/avocado-vt/sn1.qcow2"}, "driver": "qcow2"}    >>> J2

As you can see, if we translate the json string to the dict, J2 is a subset of J1.
So line 4867, we may want to check that value is a subset of v, if both value and v are dicts.
all([v[k] == value.get(k) for k in value])

Comment (YongxueHong) [reply virttest/qemu_vm.py:4867]: > Hi @YongxueHong , make sense, we can fix it in test case.
> 
> Another concern here, if we want to find an image with backing file, e.g. json:{"backing": {"driver": "qcow2", "file": {"driver": "file", "filename": "/home/kar/workspace/root/avocado/data/avocado-vt/data.qcow2"}}, "driver": "qcow2", "file": {"driver": "file", "filename": "/home/kar/workspace/root/avocado/data/avocado-vt/sn1.qcow2"}} >>> J1
> 
> but the get_image_json will not return the json string with the backing info, it just return the following: json:{"file": {"driver": "file", "filename": "/home/kar/workspace/root/avocado/data/avocado-vt/sn1.qcow2"}, "driver": "qcow2"} >>> J2
> 


Yes, your concern makes sense. Anyway, from my understanding,  we could consider whether adjust the value of returning of the API `get_image_json` if we need this requirement after researching.
Thanks.


> As you can see, if we translate the json string to the dict, J2 is a subset of J1. So line 4867, we may want to check that value is a subset of v, if both value and v are dicts. all([v[k] == value.get(k) for k in value])

**Comment by nickzhq** [comment]
@zhencliu Hello, please check this out, thanks in advance!

**Comment by nickzhq** [comment]
After discuss, the PR should be closed. All changes are on the auto-testcase

**Pull request closed without merging**